### PR TITLE
Dev adjoint double test

### PR DESCRIPTION
**Pull request by lxuechen**
Sorry that this test took so long. For long time horizons, the finite differences seem to blow up. Though apart from that particular instance, the tests pass. I also needed to use small dimensions (e.g. batch size, d, m) so that the finite differences finish under reasonable time.

### CODE REVIEW

File: tests/test_strat.py [hunk 1]
```diff
@@ -101,18 +105,113 @@ def check_efficiency():
...
+
+        swiss_knife_gradcheck(func_high_order, inputs, modules, rtol=rtol, atol=atol, eps=eps, grad_params=True)
 
 
 test_dg_ga_jvp()
 check_efficiency()
-test_adjoint()
+
+test_adjoint_inputs()
+test_adjoint_params()
```
Comment (patrick-kidger) [inline tests/test_strat.py:217]: Would it be possible to remove these calls at the bottom and just make this operate through pytest? The tests all already start with `test_` so they should be picked up. (Not sure how you want to handle `check_efficiency`, that's really more a diagnostic than a test.)

File: tests/test_strat.py [hunk 2]
```diff
@@ -101,18 +105,113 @@ def check_efficiency():
...
+                          eps: float = 1e-6,
+                          atol: float = 1e-5,
+                          rtol: float = 1e-3,
+                          grad_inputs=False,
+                          gradgrad_inputs=False,
+                          grad_params=False,
+                          gradgrad_params=False):
+    """Check grad and grad of grad wrt inputs and parameters of Modules.
+
+    Assumes func outputs a single scalar. Aimed to be as self-contained as
```
Comment (patrick-kidger) [inline tests/test_strat.py:145]: If this assumption is made, can there be an assert somewhere that validates that this is the case?

File: tests/test_strat.py [hunk 3]
```diff
@@ -101,18 +105,113 @@ def check_efficiency():
...
+    swiss_knife_gradcheck(func, y0, sde, eps=1e-7, rtol=1e-3, atol=1e-3, grad_inputs=True, gradgrad_inputs=True)
+
+
+def test_adjoint_params():
+    sde = problems.Ex5(d=d, m=m, sde_type=SDE_TYPES.stratonovich).to(device)
+    bm = BrownianInterval(t0=t0, t1=t1, shape=(batch_size, m), dtype=dtype, device=device)
+
+    def func(inputs, modules):
+        """Outputs gradient norm squared."""
+        y0, sde = inputs[0], modules[0]
```
Comment (patrick-kidger) [inline tests/test_strat.py:126]: So this is because `inputs` and `modules` are expected to be 1-tuples, right? Can there be asserts validating that this is the case?

File: tests/test_strat.py [hunk 4]
```diff
@@ -101,18 +105,113 @@ def check_efficiency():
...
+                plus_eps = func(inputs, modules)
+                flat_param[i] -= eps
+
+                flat_param[i] -= eps
+                minus_eps = func(inputs, modules)
+                flat_param[i] += eps
+
+                numerical_grad.append((plus_eps - minus_eps).detach() / (2 * eps))
+                del plus_eps, minus_eps
+        numerical_grad = torch.stack(numerical_grad)
```
Comment (patrick-kidger) [inline tests/test_strat.py:196]: Can this computation be wrapped in a `torch.no_grad()`?

File: tests/test_strat.py (hunk 1 above)
Comment (lxuechen) [reply tests/test_strat.py:217]: Done.

File: tests/test_strat.py (hunk 2 above)
Comment (lxuechen) [reply tests/test_strat.py:145]: Make it work with vector-valued functions just now. It's kinda in the style of `torch.autograd.gradgradcheck`, where a random v is used.

File: tests/test_strat.py (hunk 3 above)
Comment (lxuechen) [reply tests/test_strat.py:126]: The input format is easy to check. However, it's hard to check that the input`func` has this type of input format. 

I noted this in the docstring.

File: tests/test_strat.py (hunk 4 above)
Comment (lxuechen) [reply tests/test_strat.py:196]: I don't see that strictly necessary. Though, I updated where detach is placed.

**Comment by lxuechen** [comment]
As a side comment, I wrote this helper function mostly since it's not straightforward to test gradients wrt `nn.Parameter` using the stuff in `torch.autograd`.

**Comment by patrick-kidger** [comment]
LGTM
